reactive-banana/doc/design/design.md

```diff
@@ -56,22 +57,113 @@ The necessity of implementing `Events` as a data structure that supports updates
 
 3. *High-level*. We implement *Event* and *Behavior* in terms of `Pulse` and `Latch`. Here, we use **observable sharing** to map Haskell variables to mutable references, and to turn impure functions into pure functions where possible.
 
-### Low-level: Mutable graph
+## Low-level: Mutable graph
 
 * The difficulty of `union` in a push-based setting: Dependency order, priorities.
 * `Vault` for storage.
 * Garbage collection
 
 […]
 
-### Mid-level: Pulse and Latch
+## Mid-level: Pulse and Latch
 
 * `Build` monad.
 
 […]
 
-### High-level: Events and Behavior
-
-* Caching / observable sharing.
+## High-level: Events and Behavior
 
 […]
+
+### Events and Behaviors
+
+The final step in the implementation of the `Events` and `Behavior` types is their definition in terms of the `Pulse` and `Latch` types. The implementation is very direct, roughly `Events ~ Pulse` and `Behavior ~ Latch`, except that
+
+* we use *observable sharing* to make `Events` and `Behavior` behave more like pure values, and
+* for each `Behavior`, we keep track of a its updates, so that `Behavior a ~ (Latch a, Pulse ())`, in order to more efficiently connect it to the outside world.
+
+### Observable sharing
+
+This section explains how observable sharing is used in the implementation of `Events` and `Behavior`.
+
+*Observable sharing* is a technique that aids the implementation of *domain specific languages* (DSL) that are *embedded* in a purely functional host language. It makes the variable binding and sharing mechanism of the host language (e.g. Haskell's `let`) partially observable in the embedded language in order to avoid duplicating large subexpressions ([Claessen and Sands 1999][clsa], [Gill 2009][gill], [Kiselyov 2011][oleg],…). We use this technique for the same purpose in our implementation of FRP.
+
+To explain observable sharing, let us first consider a toy example and only then consider an FRP example. The toy example is a small language for evaluating arithmetic expressions `Exp`:
+
+    data Exp
+        = Val Int     -- plain `Int`
+        | Add Exp Exp -- addition
+        | Mul Exp Exp -- multiplication
```

Comment:
I'd suggest writing these code blocks as:

    ```haskell
    data Exp =  ...
    ```

That way we'll get syntax highlighting.